x/twap/logic_test.go

```diff
@@ -42,76 +43,287 @@ func newExpRecord(accum0, accum1 sdk.Dec) types.TwapRecord {
 	for name, test := range tests {
 		t.Run(name, func(t *testing.T) {
 			// correct expected record based off copy/paste values
-			test.expRecord.Time = test.interpolateTime
+			test.expRecord.Time = test.newTime
+			test.expRecord.PoolId = test.record.PoolId
 			test.expRecord.P0LastSpotPrice = test.record.P0LastSpotPrice
 			test.expRecord.P1LastSpotPrice = test.record.P1LastSpotPrice
 
-			gotRecord := twap.RecordWithUpdatedAccumulators(test.record, test.interpolateTime)
+			gotRecord := twap.RecordWithUpdatedAccumulators(test.record, test.newTime)
 			require.Equal(t, test.expRecord, gotRecord)
 		})
 	}
 }
 
-func (s *TestSuite) TestUpdateTwap() {
-	poolId := s.PrepareBalancerPoolWithCoins(defaultTwoAssetCoins...)
-	newSp := sdk.OneDec()
+func (s *TestSuite) TestGetInterpolatedRecord() {
+	baseRecord := newTwapRecordWithDefaults(baseTime, sdk.OneDec(), sdk.OneDec(), sdk.OneDec())
+	// tMin2Record, tMin1Record, baseRecord, tPlus1Record := s.createTestRecordsFromTime(baseTime)
 
 	tests := map[string]struct {
-		record     types.TwapRecord
-		updateTime time.Time
-		expRecord  types.TwapRecord
+		recordsToPreSet     types.TwapRecord
+		testPoolId          uint64
+		testDenom0          string
+		testDenom1          string
+		testTime            time.Time
+		expectedAccumulator sdk.Dec
+		expectedErr         bool
 	}{
-		"0 accum start": {
-			record:     newRecord(time.Unix(1, 0), sdk.NewDec(10), zeroDec, zeroDec),
-			updateTime: time.Unix(2, 0),
-			expRecord:  newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		"same time with existing record": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset0Denom,
+			testDenom1:      baseRecord.Asset1Denom,
+			testTime:        baseTime,
+		},
+		"call 1 second after existing record": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset0Denom,
+			testDenom1:      baseRecord.Asset1Denom,
+			testTime:        baseTime.Add(time.Second),
+			// 1000000000(spot price) * 1(time)
+			expectedAccumulator: baseRecord.P0ArithmeticTwapAccumulator.Add(sdk.NewDec(1000000000)),
+		},
+		"call 1 second before existing record": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset0Denom,
+			testDenom1:      baseRecord.Asset1Denom,
+			testTime:        baseTime.Add(-time.Second),
+			expectedErr:     true,
+		},
+		"on lexicographical order denom parameters": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset1Denom,
+			testDenom1:      baseRecord.Asset0Denom,
+			testTime:        baseTime,
+		},
+		"test non lexicographical order parameter": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId + 1,
+			testDenom0:      baseRecord.Asset1Denom,
+			testDenom1:      baseRecord.Asset0Denom,
+			testTime:        baseTime,
+			expectedErr:     true,
 		},
```

Comment:
Here is an example of why expectedErrors as boolean is a bad idea. This is actually erroring because the testPoolId was increased to a non existent pool, not because the denoms were in non lexicographical order (getInterpolatedRecord calls LexicographicalOrderDenoms which orders the denoms for us)

Reply:
Good catch! Agreeing with you, we should also walk towards changing expectedErr across our repo into specific errors.